network/tcp.go

```diff
@@ -0,0 +1,422 @@
+package network
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/dedis/cothority/log"
+)
+
+// NewTCPRouter returns a fresh Router using TCPHost as the underlying Host.
+func NewTCPRouter(sid *ServerIdentity) (*Router, error) {
+	h, err := NewTCPHost(sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// TCPConn is the underlying implementation of
+// Conn using plain Tcp.
+type TCPConn struct {
+	// The name of the endpoint we are connected to.
+	endpoint Address
+
+	// The connection used
+	conn net.Conn
+
+	// closed indicator
+	closed    bool
+	closedMut sync.Mutex
+	// So we only handle one receiving packet at a time
+	receiveMutex sync.Mutex
+	// So we only handle one sending packet at a time
+	sendMutex sync.Mutex
+
+	counterSafe
+}
+
+// NewTCPConn will open a TCPConn to the given address.
+func NewTCPConn(addr Address) (*TCPConn, error) {
+	var err error
+	var conn net.Conn
+	netAddr := addr.NetworkAddress()
+	for i := 0; i < MaxRetryConnect; i++ {
+		conn, err = net.Dial("tcp", netAddr)
+		if err != nil {
+			time.Sleep(WaitRetry)
+		} else {
+			break
+		}
+		time.Sleep(WaitRetry)
+	}
+	if conn == nil {
+		return nil, fmt.Errorf("Could not connect to %s: %s", addr, err)
+	}
+	c := TCPConn{
+		endpoint: addr,
+		conn:     conn,
+	}
+	return &c, err
+}
+
+// Receive calls the receive routine to get the bytes from the connection then
+// it tries to decode the buffer. Returns the Packet with the Msg field decoded
+// or EmptyApplicationPacket and an error if something wrong occured.
+func (c *TCPConn) Receive(ctx context.Context) (nm Packet, e error) {
+	defer func() {
+		if err := recover(); err != nil {
+			e = fmt.Errorf("Error Received message: %v", err)
+			nm = EmptyApplicationPacket
+		}
+	}()
+
+	var am Packet
+	buff, err := c.receive()
+	if err != nil {
+		return EmptyApplicationPacket, err
+	}
+
+	err = am.UnmarshalBinary(buff)
+	if err != nil {
+		return EmptyApplicationPacket, fmt.Errorf("Error unmarshaling message type %s: %s", am.MsgType.String(), err.Error())
+	}
+	am.From = c.Remote()
+	return am, nil
+}
+
+// receive is responsible for getting first the size of the message, then the
+// whole message. It returns the raw message as slice of bytes.
+func (c *TCPConn) receive() ([]byte, error) {
+	c.receiveMutex.Lock()
+	defer c.receiveMutex.Unlock()
+	// First read the size
+	var total Size
+	if err := binary.Read(c.conn, globalOrder, &total); err != nil {
+		return nil, handleError(err)
+	}
+	b := make([]byte, total)
+	var read Size
+	var buffer bytes.Buffer
+	for read < total {
```

Comment:
Definitely! This method should be fully covered and all edge-cases/errors should have small tests...